### DISCUSSION TITLE
Validate certificate against CRL in openssl30 vs OpenSSL 102

### DISCUSSION THREAD

**Discussion by ChauhanPriyanka** [openssl/openssl#22426] [Q&A]
I have two questions:

1. Could you please confirm if the verification is correct in OpenSSL30 is correct

```
// code looks for serial number of certificate in CRL. If found make sure reason is not removeFromCRL.
//OpenSSL 102
(rev->reason == CRL_REASON_REMOVE_FROM_CRL)
{
      return 2;
}*/

//OpenSSL 30
ASN1_INTEGER* asnSerial = NULL;
if (X509_CRL_get0_by_serial(crl, &rev, asnSerial) == CRL_REASON_REMOVE_FROM_CRL) //To be verified
{
   return 2;
}
```

2. Earlier the CRL structure in OpenSSL 102 has flags attribute, however I'm not able to find the alternate for crl->flag in openssl30. Please help what is the alternate.

**Comment by mattcaswell** [comment]
> X509_CRL_get0_by_serial(crl, &rev, asnSerial) == CRL_REASON_REMOVE_FROM_CRL

The man page for `X509_CRL_get0_by_serial` says this:

> X509_CRL_get0_by_serial() and X509_CRL_get0_by_cert() return 0 for failure,
1 on success except if the revoked entry has the reason C\<removeFromCRL\> (8),
in which case 2 is returned.

So, your code sample is not quite correct here. You should compare against the value 2 not `CRL_REASON_REMOVE_FROM_CRL`

> Earlier the CRL structure in OpenSSL 102 has flags attribute, however I'm not able to find the alternate for crl->flag in openssl30. Please help what is the alternate.

The flags are considered internal to the CRL, so you are not supposed to access them. What did you need to access the flags for?

**Comment by ChauhanPriyanka** [comment]
hi @mattcaswell,
Thank you for the response. 
As per suggestion, I will modify code.

As per the value defined in x509v3.h
```
# define CRL_REASON_NONE                         -1
# define CRL_REASON_UNSPECIFIED                  0
# define CRL_REASON_KEY_COMPROMISE               1
# define CRL_REASON_CA_COMPROMISE                2
# define CRL_REASON_AFFILIATION_CHANGED          3
# define CRL_REASON_SUPERSEDED                   4
# define CRL_REASON_CESSATION_OF_OPERATION       5
# define CRL_REASON_CERTIFICATE_HOLD             6
# define CRL_REASON_REMOVE_FROM_CRL              8
# define CRL_REASON_PRIVILEGE_WITHDRAWN          9
# define CRL_REASON_AA_COMPROMISE                10
```

```
//I have replaced CRL_REASON_REMOVE_FROM_CRL  with CRL_REASON_CA_COMPROMISE 
ASN1_INTEGER* asnSerial = NULL;
if (X509_CRL_get0_by_serial(crl, &rev, asnSerial) ==  CRL_REASON_CA_COMPROMISE ) 
{
   return 2; //internal failure return code
}

```
In openssl102 this is how the crl->flags is used. I'm trying to understand if this is applicable to openssl30
```
if (crl->flags & EXFLAG_CRITICAL) - 
	{
		X509_VERIFY_PARAM* paramCtx;
		paramCtx = X509_STORE_get0_param(ctx);
		unsigned long long flags = X509_VERIFY_PARAM_get_flags(paramCtx);
		if (flags & X509_V_FLAG_IGNORE_CRITICAL)
		{
			return 1; //success
		}
	}
```
